commit 6a6155f664e31c9be43cd690541a9a682ba3dc22
Author: George Popescu <georgepope@android.com>
Date:   Thu Oct 15 20:13:38 2020 -0700

    ubsan: introduce CONFIG_UBSAN_LOCAL_BOUNDS for Clang
    
    When the kernel is compiled with Clang, -fsanitize=bounds expands to
    -fsanitize=array-bounds and -fsanitize=local-bounds.
    
    Enabling -fsanitize=local-bounds with Clang has the unfortunate
    side-effect of inserting traps; this goes back to its original intent,
    which was as a hardening and not a debugging feature [1].  The same
    feature made its way into -fsanitize=bounds, but the traps remained.  For
    that reason, -fsanitize=bounds was split into 'array-bounds' and
    'local-bounds' [2].
    
    Since 'local-bounds' doesn't behave like a normal sanitizer, enable it
    with Clang only if trapping behaviour was requested by
    CONFIG_UBSAN_TRAP=y.
    
    Add the UBSAN_BOUNDS_LOCAL config to Kconfig.ubsan to enable the
    'local-bounds' option by default when UBSAN_TRAP is enabled.
    
    [1] http://lists.llvm.org/pipermail/llvm-dev/2012-May/049972.html
    [2] http://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20131021/091536.html
    
    Suggested-by: Marco Elver <elver@google.com>
    Signed-off-by: George Popescu <georgepope@android.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Reviewed-by: David Brazdil <dbrazdil@google.com>
    Reviewed-by: Marco Elver <elver@google.com>
    Cc: Masahiro Yamada <masahiroy@kernel.org>
    Cc: Michal Marek <michal.lkml@markovi.net>
    Cc: Nathan Chancellor <natechancellor@gmail.com>
    Cc: Nick Desaulniers <ndesaulniers@google.com>
    Cc: Kees Cook <keescook@chromium.org>
    Cc: Dmitry Vyukov <dvyukov@google.com>
    Cc: Arnd Bergmann <arnd@arndb.de>
    Cc: Peter Zijlstra <a.p.zijlstra@chello.nl>
    Link: https://lkml.kernel.org/r/20200922074330.2549523-1-georgepope@google.com
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

diff --git a/scripts/Makefile.ubsan b/scripts/Makefile.ubsan
index 27348029b2b8..4e3fff0745e8 100644
--- a/scripts/Makefile.ubsan
+++ b/scripts/Makefile.ubsan
@@ -4,7 +4,15 @@ ifdef CONFIG_UBSAN_ALIGNMENT
 endif
 
 ifdef CONFIG_UBSAN_BOUNDS
-      CFLAGS_UBSAN += $(call cc-option, -fsanitize=bounds)
+      ifdef CONFIG_CC_IS_CLANG
+            CFLAGS_UBSAN += -fsanitize=array-bounds
+      else
+            CFLAGS_UBSAN += $(call cc-option, -fsanitize=bounds)
+      endif
+endif
+
+ifdef CONFIG_UBSAN_LOCAL_BOUNDS
+      CFLAGS_UBSAN += -fsanitize=local-bounds
 endif
 
 ifdef CONFIG_UBSAN_MISC